src/main/java/org/opensearch/ad/ml/CheckpointDao.java

```diff
@@ -250,122 +224,37 @@ private void saveModelCheckpointAsync(Map<String, Object> source, String modelId
-                            // It is possible the index has been created while we sending the create request
-                            flush(bulkRequest);
-                        } else {
-                            logger.error(String.format(Locale.ROOT, "Unexpected error creating index %s", indexName), exception);
-                        }
-                    }));
-                }
-            }
-        } finally {
-            if (lock.isHeldByCurrentThread()) {
-                lock.unlock();
-            }
-        }
-    }
-
-    private void flush(BulkRequest bulkRequest) {
-        clientUtil.<BulkRequest, BulkResponse>execute(BulkAction.INSTANCE, bulkRequest, ActionListener.wrap(r -> {
-            if (r.hasFailures()) {
-                requests.addAll(BulkUtil.getIndexRequestToRetry(bulkRequest, r));
-            } else if (requests.size() >= maxBulkRequestSize / 2) {
-                // during maintenance, we may have much more waiting in the queue.
-                // trigger another flush if that's the case.
-                flush();
-            }
-        }, e -> {
-            logger.error("Failed bulking checkpoints", e);
-            // retry during next bulk.
-            for (DocWriteRequest<?> req : bulkRequest.requests()) {
-                requests.add(req);
-            }
-        }));
-    }
-
-    /**
-     * Prepare bulking the input model state to the checkpoint index.
-     * We don't save checkpoints within checkpointInterval again.
-     * @param modelState Model state
-     * @param modelId Model Id
-     */
-    public void write(ModelState<EntityModel> modelState, String modelId) {
-        write(modelState, modelId, false);
-    }
-
-    /**
-     * Prepare bulking the input model state to the checkpoint index.
-     * We don't save checkpoints within checkpointInterval again, except this
-     * is from cold start. This method will update the input state's last
-     *  checkpoint time if the checkpoint is staged (ready to be written in the
-     *  next batch).
-     * @param modelState Model state
-     * @param modelId Model Id
-     * @param coldStart whether the checkpoint comes from cold start
+     * Prepare for index request using the contents of the given model state
+     * @param modelState an entity model state
+     * @return serialized JSON map or empty map if the state is too bloated
+     * @throws IOException  when serialization fails
      */
-    public void write(ModelState<EntityModel> modelState, String modelId, boolean coldStart) {
-        Instant instant = modelState.getLastCheckpointTime();
-        // Instant.MIN is the default value. We don't save until we are sure.
-        if ((instant == Instant.MIN || instant.plus(checkpointInterval).isAfter(clock.instant())) && !coldStart) {
-            return;
+    public Map<String, Object> toIndexSource(ModelState<EntityModel> modelState) throws IOException {
```

Comment:
Why does this need to be public?

Reply:
CheckpointWriterWorker needs to call it :)

Reply:
Right, but I am a little confused about the pattern here. Shouldn't CheckpointDao perform all checkpoint reads and writes? Why would the index source be needed outside of CheckpointDao.

Reply:
Index source (the Map<String, Object> returned by toIndexSource) is part of the IndexRequest.  I need it to create an IndexRequest.  Are you suggesting I put creating index request inside CheckpointDao and CheckpointWriteWorker only get an IndexRequest from CheckpointDao?  CheckpointDao is a utility class that provides functions to deal with specifics of checkpoints.  CheckpointWriteWorkers invokes these functions to maintain a queue.  Are you suggesting I combine CheckpointWriterWorker and CheckpointDao?

Reply:
CheckpointWriterWorker only uses this function, `toIndexSource`, from CheckpointDao. It does not use any other function like [putModelCheckpoint](https://github.com/opensearch-project/anomaly-detection/pull/69/files#diff-ec6661209db81838626b3bc8ae378b8b20a7ab9e44ae6fce2d5c342ea498184aR181). 

So it seems to me that the DAO pattern is broken - checkpoint writes are happening outside of CheckpointDao. However, if you are okay with this, I am fine with it as well.

I understand that in `executeBatchRequest` the bulk request is a generic bulk request, so maybe it does not make sense to include in CheckpointDao, unless you can confirm it is a BulkRequest for checkpoint writes.

Reply:
CheckpointWriterWorker can be viewed as a Dao for HCAD to write to the checkpoint index, similar to CheckpointReadWorker. I can put these two classes to CheckpointDao, but CheckpointDao would be so complex.  I don't know how much benefit we would gain.  It violates https://en.wikipedia.org/wiki/KISS_principle

Is there any place saying we should only have one Dao class for a single database responsibility?

Reply:
No, what I was saying is CheckpointWriteWorker already has a CheckpointDao as a member but it does not use it to interact with checkpoint persistence layer. It uses it as more of a utility. I was wondering if CheckpointWriteWorker should use the checkpointDao it already has to make calls to the checkpoint index.

From my understanding, the problem with using checkpointDao for all checkpoint operations inside CheckpointWriteWorker is that it does not really fit well in BatchWorker. `executeBatchRequest` takes and executes a bulk request that writes the checkpoint. If you were to delegate writing that bulkRequest to CheckpointDao, you would have to validate in CheckpointDao that the requests inside bulkRequest are actually CheckpointWriteRequests, or else checkpointDao could be used improperly. This validation is redundant.

On [Wikipedia](https://en.wikipedia.org/wiki/Data_access_object), they say the main benefit is the isolation between business logic and persistence layer. I think the benefit of having one class for Dao for a single database responsibility is that when changes need to be made with regards to persistence, only one class needs to be changed, supporting the single responsibility principle.

All of that being said, I am okay with this implementation.

Reply:
Got it.

I moved the read/write inside Worker classes to CheckpointDao.  Please check.